### PR TITLE
Bounded proxy buffering for protocol-mode execution (Step 2.5.3)

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Updates to explicitly implement bounded buffering for protocol-mode exec, enforce stdout-purity by design, and extend tests, docs, and ExecPlan coverage. Includes lifecycle stream-purity tests and regression coverage for zero-prefix/zero-suffix stdout around proxied protocol data. ExecPlan documentation for Step 2.5.3 added and roadmap updated accordingly.
- Tests now cover lifecycle stream purity and error paths, and protocol proxy behavior under varied daemon output scenarios.
- Documentation updates reflect the bounded buffering policy and its effects on protocol-mode execution.
- The ExecPlan document is located at `docs/execplans/2-5-3-keep-proxy-buffering-bounded.md`.

## Changes
- New file: docs/execplans/2-5-3-keep-proxy-buffering-bounded.md
- Core changes to implement bounded buffering for protocol-mode exec:
  - Introduce constants for explicit buffering:
    - STDIN_BUFFER_CAPACITY
    - OUTPUT_BUFFER_CAPACITY
  - Introduce PROTOCOL_OUTPUT_CAPACITY and apply it to Bollard's `output_capacity` for protocol-mode exec.
  - Update protocol proxy stdin forwarding to use a bounded BufReader and bounded BufWriter on the input path.
  - Update protocol proxy to enforce stdout-purity contract by design (documented in module docs) and tests.
- Code changes:
  - src/engine/connection/exec/mod.rs: Added PROTOCOL_OUTPUT_CAPACITY and wired it into StartExecOptions for protocol mode; keep other modes unchanged.
  - src/engine/connection/exec/protocol.rs: Added STDIN_BUFFER_CAPACITY and OUTPUT_BUFFER_CAPACITY constants; wrap host stdin and container input with BufReader/BufWriter using these capacities; adjust forward path accordingly. Expanded module-level doc comment to state stdout-purity contract.
- Tests:
  - src/engine/connection/exec/tests/protocol_helpers.rs: Updated tests to expect `output_capacity` of Some(65_536) for protocol mode.
  - src/engine/connection/exec/tests/protocol_proxy_bdd.rs: Extended with lifecycle-purity and error-purity test scaffolding.
  - src/engine/connection/exec/tests/proxy_helpers/lifecycle_purity.rs: New lifecycle purity tests covering startup, steady-state, shutdown, and error paths.
  - tests/bdd_interactive_exec_helpers/steps.rs: Updated to expect protocol mode `output_capacity` of Some(65_536).
  - tests/features/protocol_proxy.feature: Added scenarios for lifecycle purity and purity on error paths.
- Documentation:
  - docs/podbot-design.md: Added a section describing bounded buffering implementation for protocol-mode execution (stdin/out buffers and per-chunk output capacity).
  - docs/users-guide.md: Updated to mention bounded buffering in protocol mode and the backpressure semantics.
  - docs/podbot-roadmap.md: Updated Step 2.5 checkboxes to reflect completion after gates pass.
- Other:
  - Ensure all tests and gates reflect 64 KiB buffering values and updated expectations.

### Stage A: Make stdin forwarding buffer explicitly bounded
- Add STDIN_BUFFER_CAPACITY constant and wrap host_stdin in a bounded BufReader before copy.
- Wrap container input writer in a bounded BufWriter with explicit capacity.
- Status: COMPLETE

### Stage B: Set Bollard output_capacity for protocol-mode exec
- Add PROTOCOL_OUTPUT_CAPACITY constant and set StartExecOptions.output_capacity for `ExecMode::Protocol` to `Some(PROTOCOL_OUTPUT_CAPACITY)`; keep other modes unchanged.
- Update tests that assert options to reflect the new capacity.
- Status: COMPLETE

### Stage C: Enforce stdout purity at protocol proxy seam
- Document stdout-purity contract in `src/engine/connection/exec/protocol.rs` module doc comment; ensure routing prevents non-protocol bytes from reaching host stdout.
- Status: COMPLETE

### Stage D: Add lifecycle stream-purity tests
- Add unit tests under `src/engine/connection/exec/tests/proxy_helpers/lifecycle_purity.rs` covering startup, steady-state, shutdown, and error paths.
- Status: COMPLETE

### Stage E: Add regression test for zero stdout bytes before/after proxied bytes
- Extend tests to assert no pre/post bytes around proxied output.
- Status: COMPLETE

### Stage F: Documentation and roadmap updates
- Update `docs/podbot-design.md` and `docs/users-guide.md` to reflect bounded buffering behavior and user-facing expectations.
- Mark Step 2.5 roadmap checkboxes done after gates pass.
- Status: COMPLETE

## Validation and acceptance
- STDIN_BUFFER_CAPACITY, OUTPUT_BUFFER_CAPACITY, and PROTOCOL_OUTPUT_CAPACITY are defined and used;
- Bollard output_capacity is set for protocol-mode exec;
- Lifecycle and regression tests exist and pass;
- A module-level stdout-purity doc comment exists in protocol.rs;
- Design and user documentation are updated;
- Roadmap checkboxes are updated after gates pass;
- Full gate stack passes.

## Testing plan
- Run gates: make check-fmt, make lint, make test; ensure all pass.

## Documentation
- ExecPlan 2-5-3 documented in `docs/execplans/2-5-3-keep-proxy-buffering-bounded.md`.
- Podbot design and user guide updated to reflect bounded buffering and stdout-purity.

## Notes
- This PR includes the ExecPlan document and substantive changes to implement explicit bounded buffering, tests to prove stream-purity across lifecycle stages, and user-facing documentation updates. The plan remains to enforce clear backpressure semantics and maintain protocol-mode purity at the protocol proxy seam.

- Task: https://www.devboxer.com/task/7360e9a2-744d-4153-b769-e143a7027b55

## Summary by Sourcery

Enforce bounded buffering and stdout purity for protocol-mode execution while updating tests and documentation to reflect the new behaviour.

New Features:
- Add explicit 64 KiB buffer capacities for stdin forwarding and protocol output chunks.
- Introduce lifecycle and error-path BDD scenarios for protocol proxy stream purity.

Bug Fixes:
- Ensure protocol proxy does not leak error or non-protocol bytes to host stdout when daemon stream errors occur.

Enhancements:
- Refine protocol proxy tests and helpers to capture daemon stream errors and assert exact stdout contents without prefixes or suffixes.
- Document bounded buffering implementation and stdout-purity contract for protocol-mode execution in the design docs and ExecPlan.

Documentation:
- ExecPlan 2-5-3 documents bounded proxy buffering and stream-purity enforcement; Step 2.5 marked complete in roadmap.
- Update design and user guides to describe the 64 KiB bounded buffering policy and backpressure semantics for protocol mode.

Tests:
- Add lifecycle stream-purity unit tests for startup, steady-state, shutdown, and error paths plus regression test for zero extra stdout bytes around proxied output.
- Extend protocol proxy BDD steps and feature scenarios to cover lifecycle purity and daemon stream error behaviour.
- Update existing exec tests to expect a 64 KiB protocol-mode output_capacity from Bollard.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[src/engine/connection/exec/tests/protocol_proxy_bdd.rs](https://github.com/leynos/podbot/pull/72/files/52b66f41cfbb6d3e756bcd5352b244543e9b5aa8#diff-53118f4c052a3bcdbe04e6d93188ee187c8f2e743cbf34e81ba1721a29557d9e)

Comment on file

~~~rust
        .fail_stdout_write
        .get()
        .unwrap_or(false);
    let stream_error = protocol_proxy_state.stream_error.get().unwrap_or(false);
~~~

❌ New issue: [Large Method](https://codescene.io/projects/75835/delta?repo-id=1423864&review-id=72&biomarker=Large+Method&filename=src%2Fengine%2Fconnection%2Fexec%2Ftests%2Fprotocol_proxy_bdd.rs&method=the_protocol_proxy_runs)
the_protocol_proxy_runs has 76 lines, threshold = 70

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

## Large Method
src/engine/connection/exec/tests/protocol_proxy_bdd.rs: the_protocol_proxy_runs

### What lead to degradation?
the_protocol_proxy_runs has 76 lines, threshold = 70

### Why does this problem occur?
Overly long functions make the code harder to read. The recommended maximum function length for the Rust language is 70 lines of code. Severity: Brain Method - Complex Method - Long Method.

### How to fix it?
We recommend to be careful here -- just splitting long functions don't necessarily make the code easier to read. Instead, look for natural chunks inside the functions that expresses a specific task or concern. Often, such concerns are indicated by a Code Comment followed by an if-statement. Use the [EXTRACT FUNCTION](https://refactoring.com/catalog/extractFunction.html) refactoring to encapsulate that concern.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Overall Comments
- The `OUTPUT_BUFFER_CAPACITY` constant in `protocol.rs` is used for the container stdin writer, which can be confusing given the name and neighbouring comments; consider renaming it (e.g. to `STDIN_WRITE_BUFFER_CAPACITY` or similar) to make its role clearer at the call site.
- The BDD step `host stdout contains no prefix or suffix bytes` currently only asserts a successful outcome, relying on the previous step to check content; if you want this step to stand on its own as a regression guard, consider explicitly re-checking that the captured stdout exactly matches the expected concatenated bytes.

## Individual Comments

### Comment 1
<location path="src/engine/connection/exec/mod.rs" line_range="29" />
<code_context>
+/// overhead compared to the default 8 KiB for large protocol messages. A 64 KiB
+/// capacity matches common JSON-RPC frame buffer sizes and aligns with the
+/// stdin/output buffer capacities used by the protocol proxy.
+const PROTOCOL_OUTPUT_CAPACITY: usize = 65_536;
+
 /// Boxed future type returned by [`ContainerExecClient::create_exec`].
</code_context>
<issue_to_address>
**suggestion:** Avoid duplicating the 64 KiB capacity literal in multiple places to prevent configuration drift.

This is set to 65_536 to match `STDIN_BUFFER_CAPACITY` and `OUTPUT_BUFFER_CAPACITY` in `protocol.rs`. If these are intended to remain aligned for protocol-mode sessions, consider defining a shared constant (or deriving one from another) so changing one buffer size doesn’t accidentally desync the others.

Suggested implementation:

```rust
use crate::engine::connection::protocol;

/// Maximum bytes per `LogOutput` chunk from the daemon for protocol-mode exec
/// sessions. This bounds the per-chunk memory consumption while reducing
/// overhead compared to the default 8 KiB for large protocol messages. A 64 KiB
/// capacity matches common JSON-RPC frame buffer sizes and aligns with the
/// stdin/output buffer capacities used by the protocol proxy.
const PROTOCOL_OUTPUT_CAPACITY: usize = protocol::OUTPUT_BUFFER_CAPACITY;

```

To make this compile and keep things consistent:

1. In `src/engine/connection/protocol.rs`, ensure `OUTPUT_BUFFER_CAPACITY` (and/or `STDIN_BUFFER_CAPACITY`) is declared `pub const` so it can be referenced from `exec::mod`:
   - `const OUTPUT_BUFFER_CAPACITY: usize = 65_536;` → `pub const OUTPUT_BUFFER_CAPACITY: usize = 65_536;`
2. If `protocol` is not already a sibling module of `exec` under `engine::connection`, adjust the import path accordingly (for example, `use crate::engine::connection::protocol;` may need to be `use crate::engine::connection::protocol_mode as protocol;` or similar to match your actual module name).
3. If your project’s style discourages `use` in the middle of a module, move the added `use crate::engine::connection::protocol;` to the existing `use` block at the top of `mod.rs` and reference `protocol::OUTPUT_BUFFER_CAPACITY` from there.
</issue_to_address>

### Comment 2
<location path="src/engine/connection/exec/tests/protocol_proxy_bdd.rs" line_range="300-315" />
<code_context>
+    )
+}
+
+#[then("host stdout contains no prefix or suffix bytes")]
+fn host_stdout_contains_no_extra_bytes(
+    protocol_proxy_state: &ProtocolProxyState,
+) -> StepResult<()> {
+    // This step verifies the earlier assertion covered the complete output
+    let outcome = protocol_proxy_state
+        .outcome
+        .get()
+        .ok_or_else(|| String::from("proxy outcome should be recorded"))?;
+    match outcome {
+        ProtocolProxyOutcome::Success => Ok(()),
+        ProtocolProxyOutcome::Failure(message) => Err(format!(
+            "expected successful proxy run, got failure: {message}"
+        )),
</code_context>
<issue_to_address>
**suggestion (testing):** This step doesn’t actually assert that stdout has no extra bytes; it only checks that the proxy run succeeded.

The step name implies we assert there are no prefix/suffix bytes, but this step only checks `ProtocolProxyOutcome` and never re-examines `host_stdout`. It implicitly relies on earlier steps and on `assert_channel_receives` using exact equality. To make the “no extra bytes” guarantee explicit and resilient to future changes, consider also validating here that the captured stdout exactly matches the expected concatenated bytes (including a length check), e.g., by re-reading `host_stdout` or by storing the expected bytes in shared state and checking equality/length in this step.

```suggestion
#[then("host stdout contains no prefix or suffix bytes")]
fn host_stdout_contains_no_extra_bytes(
    protocol_proxy_state: &ProtocolProxyState,
) -> StepResult<()> {
    // First, ensure the proxy run itself was successful.
    let outcome = protocol_proxy_state
        .outcome
        .get()
        .ok_or_else(|| String::from("proxy outcome should be recorded"))?;
    if let ProtocolProxyOutcome::Failure(message) = outcome {
        return Err(format!(
            "expected successful proxy run, got failure: {message}"
        ));
    }

    // Then, explicitly assert that there are no additional bytes on host stdout
    // beyond what earlier steps have already asserted.
    assert_channel_receives(
        &[],
        protocol_proxy_state.host_stdout.get(),
        "host stdout",
    )?;

    Ok(())
}
```
</issue_to_address>

### Comment 3
<location path="src/engine/connection/exec/tests/proxy_helpers/lifecycle_purity.rs" line_range="39-42" />
<code_context>
+
+use super::*;
+
+/// Startup purity: protocol proxy delivers exactly the container output with
+/// no prefix bytes from session setup.
+#[rstest]
+fn startup_purity_no_prefix_bytes(runtime: RuntimeFixture) {
+    let output_chunks = vec![Ok(LogOutput::StdOut {
+        message: b"STARTUP_OUTPUT".to_vec().into(),
+    })];
+    let output = make_output_stream(output_chunks);
+
+    let host_stdout = RecordingWriter::new();
+    let captured_stdout = host_stdout.bytes.clone();
+    let result = run_session(
+        runtime,
+        b"",
+        output,
+        Box::pin(RecordingInputWriter::new()),
+        host_stdout,
+        RecordingWriter::new(),
+    );
+
+    assert!(result.is_ok(), "startup should succeed");
+    let captured = captured_stdout.lock().expect("mutex should not poison");
+    assert_eq!(
+        captured.as_slice(),
+        b"STARTUP_OUTPUT",
</code_context>
<issue_to_address>
**suggestion (testing):** Add a lifecycle purity test for the edge case where the daemon never emits stdout bytes.

Existing lifecycle tests cover various stdout patterns but not the case where the daemon never writes to stdout (only stderr or exits silently). Please add a test that uses an empty `output_chunks` sequence and asserts that `host_stdout` remains empty while `run_session` still succeeds, to lock in the “no output at all” purity guarantee and catch accidental banner/shutdown bytes on stdout.

```suggestion
#[rstest]
fn lifecycle_purity_no_stdout_bytes(runtime: RuntimeFixture) {
    // Daemon never emits stdout bytes; ensure we don't inject any banner/prefix/suffix.
    let output_chunks: Vec<Result<LogOutput, BollardError>> = Vec::new();
    let output = make_output_stream(output_chunks);

    let host_stdout = RecordingWriter::new();
    let captured_stdout = host_stdout.bytes.clone();
    let result = run_session(
        runtime,
        b"",
        output,
        Box::pin(RecordingInputWriter::new()),
        host_stdout,
        RecordingWriter::new(),
    );

    assert!(result.is_ok(), "session should succeed even with no stdout");
    let captured = captured_stdout.lock().expect("mutex should not poison");
    assert_eq!(
        captured.as_slice(),
        b"",
        "host stdout must remain empty when container never writes to stdout"
    );
}

/// Steady-state purity: protocol proxy delivers only container stdout and
/// console bytes, routing stderr separately, and suppressing stdin echoes.
#[rstest]
fn steady_state_purity_mixed_streams(runtime: RuntimeFixture) {
```
</issue_to_address>

### Comment 4
<location path="src/engine/connection/exec/tests/proxy_helpers/lifecycle_purity.rs" line_range="167-176" />
<code_context>
+/// Regression test: zero stdout bytes before the first proxied protocol byte
</code_context>
<issue_to_address>
**suggestion (testing):** Consider adding a regression test that stresses multiple stdout chunks around the buffer size to exercise the bounded buffering behaviour.

The existing regression test `regression_zero_bytes_before_first_and_after_last_proxied_byte` is useful for the single small `StdOut` case. With the new 64 KiB bounded buffering on stdin/stdout, it would be helpful to add a similar test that (1) uses multiple `StdOut` chunks whose total size exceeds 64 KiB and (2) asserts that `host_stdout` is exactly the concatenation of those chunks, with no extra or missing bytes. That would exercise the BufReader/BufWriter + copy loop under higher throughput and specifically cover behaviour at the buffer boundary.

Suggested implementation:

```rust
/// Regression test: zero stdout bytes before the first proxied protocol byte
/// and after the final proxied byte. This guards the stdout-purity contract
/// stated in the design document and prevents future code from accidentally
/// adding banners, diagnostics, or framing bytes to the protocol stdout path.
#[rstest]
fn regression_zero_bytes_before_first_and_after_last_proxied_byte(runtime: RuntimeFixture) {
    let known_output = b"PROTOCOL_OUTPUT";
    let output_chunks = vec![Ok(LogOutput::StdOut {
        message: known_output.to_vec().into(),
    })];
    let output = make_output_stream(output_chunks);

    // existing assertions and setup for this regression test continue below...
}

/// Regression test: bounded buffering of stdout preserves all bytes across
/// multiple chunks whose total size exceeds the internal buffer size (64 KiB).
/// This specifically exercises the BufReader/BufWriter + copy loop behaviour
/// at and around the buffer boundary to ensure no bytes are lost, duplicated,
/// or reordered when proxying stdout.
#[rstest]
fn regression_stdout_bounded_buffering_preserves_all_bytes(runtime: RuntimeFixture) {
    // Use a total size > 64 KiB to cross the buffer boundary. We pick an odd
    // size to avoid aligning perfectly with any internal buffer sizes.
    const TOTAL_SIZE: usize = 70 * 1024; // 70 KiB
    const CHUNK_SIZE: usize = 8 * 1024 + 123; // ~8 KiB, intentionally non-power-of-two

    let mut expected = Vec::with_capacity(TOTAL_SIZE);
    let mut output_chunks = Vec::new();

    let mut remaining = TOTAL_SIZE;
    let mut byte_value: u8 = 0;

    while remaining > 0 {
        let this_chunk = remaining.min(CHUNK_SIZE);
        let mut chunk = Vec::with_capacity(this_chunk);

        for _ in 0..this_chunk {
            // Deterministic but non-trivial pattern so that reordering or
            // duplication would be visible in the final concat.
            chunk.push(byte_value);
            expected.push(byte_value);
            byte_value = byte_value.wrapping_add(1);
        }

        output_chunks.push(Ok(LogOutput::StdOut {
            message: chunk.into(),
        }));

        remaining -= this_chunk;
    }

    let output = make_output_stream(output_chunks);

    // Run the exec path that proxies stdout from the protocol to the host.
    let (result, captured_stdout) = run_exec_with_proxied_stdout(runtime, output);

    // The exec is expected to succeed; adjust the assertion if this helper
    // reports success differently in your test harness.
    assert_exec_succeeded(&result);

    // The host stdout must exactly equal the concatenation of all stdout chunks,
    // with no extra or missing bytes.
    let captured = captured_stdout.lock().expect("mutex should not poison");
    assert_eq!(
        captured.as_slice(),
        expected.as_slice(),
        "host stdout must be exactly the concatenation of all StdOut chunks \
         even when total size exceeds the bounded buffer size"
    );

```

1. Ensure there is a helper `run_exec_with_proxied_stdout(runtime, output)` returning `(result, captured_stdout)` in this module, matching the pattern used by the existing tests (the first snippet in your message suggests such a helper already exists). If the name or signature differs, update the new test to use the correct helper.
2. Ensure `assert_exec_succeeded(&result)` exists; if your existing tests assert success differently (e.g. `assert_exec_ok(&result)` or pattern matching on `result`), update the assertion accordingly to match the conventions in this file.
3. If your `make_output_stream` or `LogOutput::StdOut` constructors have slightly different argument types (e.g. `Bytes` vs `Vec<u8>`), adjust the `message: chunk.into()` calls to the appropriate conversion method consistent with the rest of the file.
</issue_to_address>

### Comment 5
<location path="tests/bdd_interactive_exec_helpers/steps.rs" line_range="183-186" />
<code_context>
                     Some(bollard::exec::StartExecOptions {
                         detach: false,
                         tty: false,
-                        output_capacity: None
+                        output_capacity: Some(65_536)
                     })
                 );
</code_context>
<issue_to_address>
**suggestion (testing):** Consider adding/ensuring explicit coverage that non-protocol exec modes still use `output_capacity: None`.

Since this helper now expects `output_capacity: Some(65_536)` for protocol-mode exec and the code only sets `PROTOCOL_OUTPUT_CAPACITY` for `ExecMode::Protocol`, please add (or confirm) a test/BDD scenario for attached/detached exec that asserts `StartExecOptions.output_capacity` stays `None`. That will protect against future refactors accidentally enabling bounded output for non-protocol modes.

Suggested implementation:

```rust
                    Some(bollard::exec::StartExecOptions {
                        detach: false,
                        tty: false,
                        output_capacity: Some(65_536)
                    })
                );

                // Explicitly cover non-protocol exec: output_capacity must remain None
                let non_protocol_start_exec_options = bollard::exec::StartExecOptions {
                    detach: false,
                    tty: false,
                    output_capacity: None,
                };
                assert_eq!(None, non_protocol_start_exec_options.output_capacity);

                let output_stream = stream::iter(vec![Ok(LogOutput::StdOut {

```

The added check ensures that a `StartExecOptions` instance explicitly configured as non-protocol keeps `output_capacity: None`, but it does not yet exercise your actual helper/exec-mode wiring.

To fully align with the review comment, you should:
1. Identify the helper function (or BDD step) that constructs `StartExecOptions` based on `ExecMode` (e.g., `ExecMode::Attached`, `ExecMode::Detached`, `ExecMode::Protocol`).
2. Add a BDD scenario (or unit test in this file) that:
   - Invokes that helper with a non-protocol mode (attached/detached).
   - Asserts that the resulting `StartExecOptions.output_capacity` is `None`.
3. If the helper currently only has a Protocol-mode path explicitly covered in tests, mirror that structure:
   - Re-use whatever pattern you’re using to assert `output_capacity: Some(65_536)` for `ExecMode::Protocol`.
   - Add analogous expectations for `ExecMode::Attached` / `ExecMode::Detached` asserting `None`.

You’ll need to adapt these steps to the actual helper signatures and step definitions present in `steps.rs`, so the new tests go through the same code path that your production code uses for non-protocol exec.
</issue_to_address>

### Comment 6
<location path="docs/podbot-design.md" line_range="194" />
<code_context>
+  backpressure propagates to the container.
+
+The 64 KiB buffer sizes align with common protocol message sizes (such as
+JSON-RPC frame buffers) and typical OS pipe buffer defaults, keeping latency
+low while preventing unbounded accumulation during high-throughput scenarios.
+
</code_context>
<issue_to_address>
**suggestion (review_instructions):** "JSON-RPC" is introduced here without expanding the acronym, which may violate the requirement to define uncommon acronyms on first use.

Consider expanding this as "JSON Remote Procedure Call (JSON-RPC)" the first time it appears, so the acronym is defined explicitly for readers.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 7
<location path="docs/execplans/2-5-3-keep-proxy-buffering-bounded.md" line_range="105" />
<code_context>
+  protocol messages to be split into multiple `LogOutput` chunks, increasing
+  per-chunk overhead. Setting it too high defeats the backpressure goal.
+  Severity: medium. Likelihood: low. Mitigation: choose a value that matches
+  typical protocol message sizes (64 KiB is common for JSON-RPC frame
+  buffers). Document the choice in the decision log and design document so it
+  can be tuned later.
</code_context>
<issue_to_address>
**suggestion (review_instructions):** "JSON-RPC" is used here without an accompanying expansion, which may not satisfy the requirement to define uncommon acronyms on first use.

Please expand this on first use, for example: "JSON Remote Procedure Call (JSON-RPC) frame buffers". Subsequent references can then use just "JSON-RPC".

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 8
<location path="docs/execplans/2-5-3-keep-proxy-buffering-bounded.md" line_range="323" />
<code_context>
+   error-related bytes written to stdout. The error should surface as a
+   `PodbotError`.
+
+Required BDD test scenarios (add to
+`tests/features/protocol_proxy.feature`):
+
</code_context>
<issue_to_address>
**suggestion (review_instructions):** "BDD" is introduced here without being defined, which conflicts with the requirement to define uncommon acronyms on first use.

Consider expanding this as "behaviour-driven development (BDD)" at its first occurrence, so the acronym is clear to readers.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~

File: src/engine/connection/exec/protocol.rs [hunk 1]
```diff
@@ -147,15 +177,17 @@ async fn abort_stdin_forwarding_task(stdin_task: JoinHandle<io::Result<()>>) {
...
 where
     HostStdin: AsyncRead + Unpin,
 {
-    tokio::io::copy(&mut host_stdin, &mut input).await?;
-    input.flush().await?;
-    input.shutdown().await
+    let mut buffered_stdin = tokio::io::BufReader::with_capacity(STDIN_BUFFER_CAPACITY, host_stdin);
+    let mut buffered_input = tokio::io::BufWriter::with_capacity(OUTPUT_BUFFER_CAPACITY, input);
+    tokio::io::copy(&mut buffered_stdin, &mut buffered_input).await?;
+    buffered_input.flush().await?;
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/engine/connection/exec/protocol.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[src/engine/connection/exec/tests/proxy_helpers/lifecycle_purity.rs](https://github.com/leynos/podbot/pull/72/files/4ad4a47537c60d973ddb34ea28f7a4dbda13210d#diff-4946e1104f5a6ee358e81bf2838eeb529058238c7209a08eb33a8c5823a1e094)

Comment on file

~~~rust
fn startup_purity_no_prefix_bytes(runtime: RuntimeFixture) {
    let output_chunks = vec![Ok(LogOutput::StdOut {
        message: b"STARTUP_OUTPUT".to_vec().into(),
    })];
    let output = make_output_stream(output_chunks);

    let host_stdout = RecordingWriter::new();
    let captured_stdout = host_stdout.bytes.clone();
    let result = run_session(
        runtime,
        b"",
        output,
        Box::pin(RecordingInputWriter::new()),
        host_stdout,
        RecordingWriter::new(),
    );

    assert!(result.is_ok(), "startup should succeed");
    let captured = captured_stdout.lock().expect("mutex should not poison");
    assert_eq!(
        captured.as_slice(),
        b"STARTUP_OUTPUT",
        "host stdout must contain exactly the container output with no prefix"
    );
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/75835/delta?repo-id=1423864&review-id=72&biomarker=Code+Duplication&filename=src%2Fengine%2Fconnection%2Fexec%2Ftests%2Fproxy_helpers%2Flifecycle_purity.rs&method=)
The module contains 3 functions with similar structure: lifecycle_purity_no_stdout_bytes,shutdown_purity_no_suffix_bytes,startup_purity_no_prefix_bytes

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[src/engine/connection/exec/tests/proxy_helpers/lifecycle_purity.rs](https://github.com/leynos/podbot/pull/72/files/6c4e78cdf6c8e161b59222a0757433e0f761affe#diff-4946e1104f5a6ee358e81bf2838eeb529058238c7209a08eb33a8c5823a1e094)

Comment on file

~~~rust
fn startup_purity_no_prefix_bytes(runtime: RuntimeFixture) {
    run_stdout_purity_test(
        runtime,
        vec![Ok(LogOutput::StdOut {
            message: b"STARTUP_OUTPUT".to_vec().into(),
        })],
        b"STARTUP_OUTPUT",
        "startup should succeed",
        "host stdout must contain exactly the container output with no prefix",
    );
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/75835/delta?repo-id=1423864&review-id=72&biomarker=Code+Duplication&filename=src%2Fengine%2Fconnection%2Fexec%2Ftests%2Fproxy_helpers%2Flifecycle_purity.rs&method=)
The module contains 2 functions with similar structure: shutdown_purity_no_suffix_bytes,startup_purity_no_prefix_bytes

File: src/engine/connection/exec/protocol.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (leynos) [reply src/engine/connection/exec/protocol.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

[the remark this answers is not in the document]
Comment (leynos) [reply src/engine/connection/exec/protocol.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Module-Level Documentation | ❓ Inconclusive | Cannot verify module-level documentation compliance—files could not be accessed to confirm all modules contain proper docstrings. | Examine repository files to confirm all new/modified modules contain module-level docstrings (//! or /\*! \*/) before any code. |
